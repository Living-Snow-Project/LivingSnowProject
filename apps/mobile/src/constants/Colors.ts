const tintColor = "#2f95dc";

export const Colors = {
  tintColor,
  tabIconDefault: "#0077ff",
  tabIconSelected: tintColor,
  tabBar: "#fefefe",
  errorBackground: "red",
  errorText: "#fff",
  warningBackground: "#EAEB5E",
  warningText: "#666804",
  noticeBackground: tintColor,
  noticeText: "#fff",
};
