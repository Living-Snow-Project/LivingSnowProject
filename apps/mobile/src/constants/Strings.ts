// for TextInput placeholder prop
export const Placeholders = {
  GPS: {
    AcquiringLocation: "Looking for GPS signal",
    EnterCoordinates: "ie. 12.34567, -123.45678",
    NoPermissions:
      "Check location permissions in Settings. Enter coordinates manually.",
    NoLocation: "Could not determine location. Enter coordinates manually.",
  },
  Settings: {
    Username: "Enter your name",
    Organization: "Enter organization you belong to (if any)",
  },
  RecordScreen: {
    LocationDescription: "ie: Blue Lake, North Cascades, WA",
    Notes: "ie. dark red algae on glacial ice",
    Size: "Select a size",
    TubeId: "Leave blank if tube does not have an id",
    TubeIdDisabled: "Disabled - only used for Samples",
  },
};

export const Labels = {
  Date: "Date",
  Delete: "Delete",
  RecordForm: {
    Color: "Color (select all that apply)",
    Gps: "GPS Coordinates (latitude, longitude)",
    Photos: "Select Photos (limit 4)",
    Type: "Are you Reporting a Sighting or Taking a Sample?",
  },
  // TODO: rename to RecordDetailsFields
  RecordFields: {
    Type: "Type",
    Name: "Name",
    Organization: "Organization",
    TubeId: "Tube Id",
    GPSCoordinates: "Coordinates",
    Latitude: "Latitude",
    Longitude: "Longitude",
    Size: "Size",
    Colors: "Colors",
    LocationDescription: "Location Description",
    Notes: "Notes",
    Photos: "Photos",
  },
  TimelineScreen: {
    ExampleRecords: "Example Record",
    PendingRecords: "Pending",
    DownloadedRecords: "Downloaded",
  },
  SettingsScreen: {
    DarkMode: "Dark",
    ManualCoordinates: "Manual Coordinates Warning",
  },
  StatusBar: {
    NoConnection: "No Internet Connection",
  },
  Modal: {
    Cancel: "Cancel",
    Confirm: "Confirm",
    DiskUsage: {
      header: "Confirm Delete",
      body: "Are you sure you want to delete saved photos?",
    },
    GpsManualEntry: {
      header: "Enter coordinates manually?",
      body: `Some users enter coordinates manually if they cannot acquire GPS signal or when they return home.\n\nThis message can be disabled in Settings.`,
    },
  },
};

// for Modal alerts
export const Notifications = {
  uploadSuccess: {
    title: "Record Uploaded",
    message: "Thanks for your submission.",
  },
  uploadRecordFailed: {
    title: "Record Saved",
    message: "We will upload it later.",
  },
  uploadPhotoFailed: {
    title: "Photo Saved",
    message: "We will upload it later.",
  },
  uploadPhotosFailed: {
    title: "Photos Saved",
    message: "We will upload them later.",
  },
  updateRecordSuccess: {
    title: "Record Updated",
  },
  updateRecordFailed: {
    title: "Update Failed",
    message:
      "This can happen when Internet connectivity is restored while updating the record.",
  },
  backgroundTasksNotAllowed: {
    title: "Background Uploads Not Enabled",
    message:
      "Please enable Background App Refresh if you would like stored records to be uploaded in the background.",
  },
};

export const Validations = {
  invalidCoordinates:
    'Coordinates must be in "lat, long" format. eg. 12.34567, -123.45678',
  invalidAlgaeSize: "A size must be specified",
  invalidAlgaeColors: "At least 1 color must be specified",
};

export const Errors = {
  recordNotFound: "Record to update not found",
};

export const RecordDescription = {
  Sample: `Sample`,
  Sighting: `Sighting`,
  Undefined: "Undefined",
};

export const AlgaeSizeDescription = {
  Select: "Select a size",
  Fist: "Fist",
  ShoeBox: "Shoe Box",
  CoffeeTable: "Coffee Table",
  Car: "Car",
  Bus: "Bus",
  Playground: "Playground",
  SportsField: "Sports Field",
  Other: "Other (describe in notes)",
};

export const AlgaeColorDescription = {
  Select: "Select a color",
  Red: "Red",
  Pink: "Pink",
  Grey: "Grey",
  Green: "Green",
  Orange: "Orange",
  Yellow: "Yellow",
  Other: "Other (describe in notes)",
};

export const BackgroundTasks = {
  UploadData: "background-upload-task",
};

// TODO: move this to Labels
export const Headers = {
  DiskUsage: "Disk Usage",
  Profile: "Profile",
  Prompts: "Prompts",
  Theme: "Theme",
};
