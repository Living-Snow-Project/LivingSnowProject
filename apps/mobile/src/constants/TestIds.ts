export const TestIds = {
  GPS: {
    GpsManualEntryModal: "gps-enable-manual-coordinates-modal",
    GpsManualEntryPressable: "gps-manual-pressable",
  },
  Selectors: {
    AlgaeColor: "algae-color-selector",
    AlgaeSize: "algae-size-selector",
    Photos: "photo-selector",
    RecordType: "record-type-selector",
  },
  RecordScreen: {
    UploadButton: "record-screen-upload-button",
    UpdateButton: "record-screen-update-button",
  },
  TimelineScreen: {
    FlatList: "timeline-screen-flatlist",
    RecordListView: "timeline-screen-record-list-view",
    NewRecordButton: "timeline-screen-record-button",
    SettingsButton: "timeline-screen-settings-button",
    ScrollToTopButton: "timeline-screen-scroll-to-top-button",
  },
  SettingsScreen: {
    ShowGpsWarning: "settings-screen-show-gps-warning",
  },
  Icons: {
    AddPhotosIcon: "add-photos-icon",
    DeleteIcon: "delete-icon",
    DoneSelectingPhotosIcon: "done-selecting-photos-icon",
    EditIcon: "edit-icon",
    PictureIcon: "picture-icon",
    SampleIcon: "sample-icon",
    SaveIcon: "save-icon",
    ScrollTopIcon: "scroll-top-icon",
    SightingIcon: "sighting-icon",
    SnowIcon: "snow-icon",
  },
  RecordList: {
    MenuTrigger: "pending-record-menu-trigger",
    EditRecordAction: "edit-record-action",
    DeleteRecordAction: "delete-record-action",
  },
  Modal: {
    ConfirmButton: "modal-confirm-button",
    DeletePendingRecord: "delete-pending-record",
    DiskUsageDeletePhotos: "disk-usage-delete-photos-modal",
    NoButton: "modal-no-button",
  },
};
