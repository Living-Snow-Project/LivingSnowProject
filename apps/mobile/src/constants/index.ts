export * from "./Colors";
export * from "./Strings";
export * from "./TestIds";
