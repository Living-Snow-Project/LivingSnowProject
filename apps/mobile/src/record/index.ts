export * from "./Color";
export * from "./Record";
export * from "./Size";
