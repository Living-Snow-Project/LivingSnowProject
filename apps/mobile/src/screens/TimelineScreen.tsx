import React, { useEffect, useState } from "react";
import { FlatList } from "native-base";
import NetInfo from "@react-native-community/netinfo";
import Logger from "@livingsnow/logger";
import { TimelineScreenNavigationProp } from "../navigation/Routes";
import StatusBar from "../components/StatusBar";
import { ExampleRecordList } from "../components/RecordList";
import TestIds from "../constants/TestIds";
import { useAlgaeRecordsContext } from "../hooks/useAlgaeRecords";
import useRecordList from "../hooks/useRecordList";

type TimelineScreenProps = {
  navigation: TimelineScreenNavigationProp;
};

export default function TimelineScreen({ navigation }: TimelineScreenProps) {
  const recordList = useRecordList(navigation);
  const [connected, setConnected] = useState<boolean>(true);
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const algaeRecordsContext = useAlgaeRecordsContext();

  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener(({ isConnected }) => {
      setConnected(!!isConnected);

      if (isConnected) {
        setRefreshing(true);
      }
    });

    return unsubscribe;
  }, []);

  useEffect(() => {
    if (!refreshing || !connected) {
      return;
    }

    algaeRecordsContext
      .fullSync()
      .catch(() =>
        Logger.Warn(`Could not download records. Please try again later.`)
      )
      .finally(() => setRefreshing(false));
  }, [refreshing, connected]);

  const renderItem = ({ item }) => item;

  const onRefreshHandler = () => {
    if (connected) {
      setRefreshing(true);
    }
  };

  // TODO: called repeatedly when "true-end" reached and results in appending duplicate records
  const onEndReached = () => {
    // keep an eye on this (if list is "empty" and it gets called)
    const downloadedRecords = algaeRecordsContext.getDownloadedRecords();
    algaeRecordsContext.downloadNextRecords(
      downloadedRecords[downloadedRecords.length - 1].date
    );
  };

  return (
    <>
      <StatusBar
        state={algaeRecordsContext.getCurrentState()}
        isConnected={connected}
      />
      <FlatList
        data={recordList}
        testID={TestIds.TimelineScreen.FlatList}
        renderItem={renderItem}
        ListEmptyComponent={ExampleRecordList}
        onRefresh={onRefreshHandler}
        refreshing={false} // StatusBar component handles activity
        onEndReachedThreshold={5}
        onEndReached={onEndReached}
      />
    </>
  );
}
