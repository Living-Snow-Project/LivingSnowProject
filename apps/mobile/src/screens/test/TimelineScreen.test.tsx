import React from "react";
import { Alert } from "react-native";
import { act, fireEvent, render, waitFor } from "@testing-library/react-native";
import NetInfo, {
  NetInfoChangeHandler,
  NetInfoState,
} from "@react-native-community/netinfo";
import { AlgaeRecord, makeExampleRecord } from "@livingsnow/record";
import { RecordReducerState } from "../../../types/AlgaeRecords";
import {
  RootStackNavigationProp,
  TimelineScreenNavigationProp,
} from "../../navigation/Routes";
import TimelineScreen from "../TimelineScreen";
import {
  SettingsButton,
  NewRecordButton,
} from "../../navigation/MainTabNavigator";
import * as RecordManager from "../../lib/RecordManager";
import * as Storage from "../../lib/Storage";
import TestIds from "../../constants/TestIds";
import { Labels } from "../../constants/Strings";
import { mockedNavigate } from "../../../jesttest.setup";
import {
  RecordReducerActionsContext,
  RecordReducerStateContext,
} from "../../hooks/useAlgaeRecords";
import {
  makeRecordReducerActionsMock,
  makeRecordReducerStateMock,
} from "../../mocks/useRecordReducer.mock";

// TimelineScreen takes navigation input prop
const navigation = {} as TimelineScreenNavigationProp;
navigation.navigate = jest.fn();

const downloadedRecord: AlgaeRecord = {
  ...makeExampleRecord("Sample"),
  id: 1,
};

const pendingRecord: AlgaeRecord = {
  ...makeExampleRecord("Sighting"),
  id: 3,
};

const setIsConnected = (isConnected: boolean): void => {
  jest.spyOn(NetInfo, "addEventListener").mockImplementationOnce((callback) => {
    callback({ isConnected } as NetInfoState);

    return () => {};
  });
};

const setupDownloadSuccess = () => {
  const retryRecordsSpy = jest
    .spyOn(RecordManager, "retryPendingRecords")
    .mockImplementationOnce(() => Promise.resolve([]));

  const recordReducerStateMock = makeRecordReducerStateMock();

  recordReducerStateMock.downloadedRecords = [
    downloadedRecord,
    makeExampleRecord("Sighting"),
  ];

  setIsConnected(true);

  return {
    retryRecordsSpy,
    recordReducerStateMock,
  };
};

const setupDownloadFailed = () => {
  const retryRecordsSpy = jest
    .spyOn(RecordManager, "retryPendingRecords")
    .mockImplementationOnce(() => Promise.reject());

  setIsConnected(true);

  return retryRecordsSpy;
};

describe("TimelineScreen test suite", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders", () => {
    const { toJSON } = render(<TimelineScreen navigation={navigation} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("download records succeeds", async () => {
    const { retryRecordsSpy, recordReducerStateMock } = setupDownloadSuccess();

    const { getByTestId, getByText } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() => getByTestId(downloadedRecord.id.toString()));

    expect(retryRecordsSpy).toBeCalledTimes(1);
    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();
  });

  test("download records fails", async () => {
    const retryRecordsSpy = setupDownloadFailed();

    const { getByText } = render(<TimelineScreen navigation={navigation} />);

    await waitFor(() => getByText(Labels.TimelineScreen.ExampleRecords));
    expect(retryRecordsSpy).toBeCalledTimes(1);
  });

  test("connection lost", async () => {
    setIsConnected(false);
    const { getByText } = render(<TimelineScreen navigation={navigation} />);

    await waitFor(() => getByText(Labels.StatusBar.NoConnection));
    expect(getByText(Labels.StatusBar.NoConnection)).toBeTruthy();
  });

  test("connection restored", async () => {
    let cacheCallback: NetInfoChangeHandler;

    jest
      .spyOn(NetInfo, "addEventListener")
      .mockImplementationOnce((callback) => {
        cacheCallback = callback;
        callback({ isConnected: false } as NetInfoState);
        return () => {};
      });

    const { getByText, queryByText } = render(
      <TimelineScreen navigation={navigation} />
    );

    await waitFor(() => getByText(Labels.StatusBar.NoConnection));
    expect(getByText(Labels.StatusBar.NoConnection)).toBeTruthy();

    await act(async () => cacheCallback({ isConnected: true } as NetInfoState));
    expect(queryByText(Labels.StatusBar.NoConnection)).toBeFalsy();
  });

  test("pull to refresh with connection", async () => {
    const { recordReducerStateMock } = setupDownloadSuccess();

    const { getByText, getByTestId } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() =>
      getByTestId(recordReducerStateMock.downloadedRecords[0].id.toString())
    );
    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();

    await act(async () =>
      getByTestId(TestIds.TimelineScreen.FlatList).props.onRefresh()
    );

    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();
  });

  test("pull to refresh without connection", async () => {
    setIsConnected(false);

    const { getByText, getByTestId } = render(
      <TimelineScreen navigation={navigation} />
    );

    await act(async () =>
      getByTestId(TestIds.TimelineScreen.FlatList).props.onRefresh()
    );

    expect(getByText(Labels.TimelineScreen.ExampleRecords)).toBeTruthy();
  });

  test("onEndReached download records", async () => {
    /* eslint-disable react/jsx-no-constructed-context-values */
    const recordReducerStateMock = {
      ...makeRecordReducerStateMock(),
      downloadedRecords: [downloadedRecord],
    };
    const recordReducerActionsMock = makeRecordReducerActionsMock();

    const { getByText, getByTestId } = render(
      <RecordReducerActionsContext.Provider value={recordReducerActionsMock}>
        <RecordReducerStateContext.Provider value={recordReducerStateMock}>
          <TimelineScreen navigation={navigation} />
        </RecordReducerStateContext.Provider>
      </RecordReducerActionsContext.Provider>
    );

    await waitFor(() =>
      getByTestId(recordReducerStateMock.downloadedRecords[0].id.toString())
    );
    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();

    await act(async () =>
      getByTestId(TestIds.TimelineScreen.FlatList).props.onEndReached()
    );

    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();
  });

  test("record flatlist onLayout", async () => {
    const { recordReducerStateMock } = setupDownloadSuccess();

    const { getByTestId } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() =>
      getByTestId(recordReducerStateMock.downloadedRecords[0].id.toString())
    );

    fireEvent(getByTestId(TestIds.TimelineScreen.RecordListView), "onLayout", {
      nativeEvent: {
        layout: {
          width: 600,
          height: 1000,
        },
      },
    });
  });

  test("onScroll shows scrollToTop", async () => {
    const { recordReducerStateMock } = setupDownloadSuccess();

    const { getByTestId } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() =>
      getByTestId(recordReducerStateMock.downloadedRecords[0].id.toString())
    );

    await act(async () =>
      getByTestId(TestIds.TimelineScreen.FlatList).props.onScroll({
        nativeEvent: {
          contentSize: {
            height: 1200,
          },
          layoutMeasurement: {
            height: 1200,
          },
          contentOffset: {
            y: 1001,
          },
        },
      })
    );

    expect(getByTestId(TestIds.TimelineScreen.ScrollToTopButton)).toBeTruthy();
  });

  test("onScroll doesn't show scrollToTop", async () => {
    const { recordReducerStateMock } = setupDownloadSuccess();

    const { getByTestId, queryByTestId } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() =>
      getByTestId(recordReducerStateMock.downloadedRecords[0].id.toString())
    );

    await act(async () =>
      getByTestId(TestIds.TimelineScreen.FlatList).props.onScroll({
        nativeEvent: {
          contentSize: {
            height: 1200,
          },
          layoutMeasurement: {
            height: 1200,
          },
          contentOffset: {
            y: 999,
          },
        },
      })
    );

    expect(queryByTestId(TestIds.TimelineScreen.ScrollToTopButton)).toBeFalsy();
  });

  test("scrollToTop", async () => {
    const { recordReducerStateMock } = setupDownloadSuccess();

    const { getByTestId } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() =>
      getByTestId(recordReducerStateMock.downloadedRecords[0].id.toString())
    );

    await act(async () =>
      getByTestId(TestIds.TimelineScreen.FlatList).props.onScroll({
        nativeEvent: {
          contentSize: {
            height: 1200,
          },
          layoutMeasurement: {
            height: 1200,
          },
          contentOffset: {
            y: 1001,
          },
        },
      })
    );

    fireEvent.press(getByTestId(TestIds.TimelineScreen.ScrollToTopButton));
  });

  test("navigate to SettingsScreen", () => {
    const navigationInner = {} as RootStackNavigationProp;
    navigationInner.navigate = jest.fn();

    const { getByTestId } = render(
      <SettingsButton navigation={navigationInner} />
    );
    fireEvent.press(getByTestId(TestIds.TimelineScreen.SettingsButton));
    expect(navigationInner.navigate).toBeCalledWith("Settings");
  });

  test("navigate to RecordScreen", async () => {
    const navigationInner = {} as RootStackNavigationProp;
    navigationInner.navigate = jest.fn();

    const { getByTestId } = render(
      <NewRecordButton navigation={navigationInner} />
    );
    fireEvent.press(getByTestId(TestIds.TimelineScreen.NewRecordButton));
    expect(navigationInner.navigate).toBeCalledWith("Record");
  });

  test("navigate to record details screen", async () => {
    const { retryRecordsSpy, recordReducerStateMock } = setupDownloadSuccess();

    const { getByTestId } = render(
      <RecordReducerStateContext.Provider value={recordReducerStateMock}>
        <TimelineScreen navigation={navigation} />
      </RecordReducerStateContext.Provider>
    );

    await waitFor(() => getByTestId(downloadedRecord.id.toString()));

    fireEvent.press(getByTestId(downloadedRecord.id.toString()));

    expect(retryRecordsSpy).toBeCalledTimes(1);
    expect(mockedNavigate).toBeCalledWith("RecordDetails", {
      record: downloadedRecord,
    });
  });

  describe("Pending records tests", () => {
    const recordReducerActionsMock = makeRecordReducerActionsMock();
    const customRender = () => {
      // comes with a single pending record
      const recordReducerStateMock = makeRecordReducerStateMock();
      const renderResult = render(
        <RecordReducerStateContext.Provider value={recordReducerStateMock}>
          <RecordReducerActionsContext.Provider
            value={recordReducerActionsMock}
          >
            <TimelineScreen navigation={navigation} />
          </RecordReducerActionsContext.Provider>
        </RecordReducerStateContext.Provider>
      );

      return {
        renderResult,
        recordReducerStateMock,
      };
    };

    test("pending records render", async () => {
      const retryRecordsSpy = setupDownloadFailed();

      /* eslint-disable react/jsx-no-constructed-context-values */
      const recordReducerStateMock: RecordReducerState = {
        ...makeRecordReducerStateMock(),
        pendingRecords: [pendingRecord],
      };

      const { getByTestId, getByText } = render(
        <RecordReducerStateContext.Provider value={recordReducerStateMock}>
          <TimelineScreen navigation={navigation} />
        </RecordReducerStateContext.Provider>
      );

      // visible
      await waitFor(() => getByTestId(pendingRecord.id.toString()));
      expect(retryRecordsSpy).toBeCalledTimes(1);
      expect(getByText(Labels.TimelineScreen.PendingRecords)).toBeTruthy();
    });

    test("delete pending record", () => {
      // validates a bug where pending records were not displayed when not connected
      setIsConnected(false);
      let okPressed = false;
      const alertMock = jest
        .spyOn(Alert, "alert")
        .mockImplementationOnce((title, message, buttons) => {
          // @ts-ignore
          buttons[0].onPress(); // "Yes" button
          okPressed = true;
        });

      const { getByTestId, getByText } = customRender().renderResult;

      expect(getByText(Labels.TimelineScreen.PendingRecords)).toBeTruthy();

      fireEvent.press(getByTestId(TestIds.RecordList.DeleteRecordAction));

      return waitFor(() => expect(okPressed).toBe(true)).then(() =>
        Storage.loadPendingRecords().then((received) => {
          expect(received).toEqual([]);
          expect(alertMock).toBeCalledTimes(1);
        })
      );
    });

    test("cancel delete record", () => {
      const noButtonMock = jest.fn();
      const alertMock = jest
        .spyOn(Alert, "alert")
        .mockImplementationOnce((title, message, buttons) => {
          // @ts-ignore
          noButtonMock(buttons[1]); // "No" button
        });

      const { getByTestId } = customRender().renderResult;

      fireEvent.press(getByTestId(TestIds.RecordList.DeleteRecordAction));
      expect(alertMock).toBeCalledTimes(1);
      expect(noButtonMock).toHaveBeenLastCalledWith({
        onPress: expect.any(Function),
        text: "No",
        style: "cancel",
      });
    });

    test("edit record", () => {
      const { recordReducerStateMock, renderResult } = customRender();

      fireEvent.press(
        renderResult.getByTestId(TestIds.RecordList.EditRecordAction)
      );

      expect(navigation.navigate).toBeCalledWith("Record", {
        record: recordReducerStateMock.pendingRecords[0],
      });
    });
  });
});
