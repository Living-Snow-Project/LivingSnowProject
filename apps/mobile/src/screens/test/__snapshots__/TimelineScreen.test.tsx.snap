// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineScreen test suite renders 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <Text
    dataSet={{}}
    style={
      {
        "backgroundColor": undefined,
        "color": "#171717",
        "fontFamily": undefined,
        "fontSize": 14,
        "fontStyle": "normal",
        "fontWeight": "600",
        "letterSpacing": 0,
        "lineHeight": 21,
        "textAlign": "center",
        "textDecorationLine": undefined,
      }
    }
  >
    Idle
  </Text>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      [
        {},
        {
          "dataSet": {},
        },
        {},
      ]
    }
    data={[]}
    dataSet={{}}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={{}}
    testID="timeline-screen-flatlist"
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View>
      <View
        dataSet={{}}
        style={
          {
            "backgroundColor": "#34d399",
            "width": "100%",
          }
        }
      >
        <Text
          dataSet={{}}
          style={
            {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "600",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textAlign": "center",
              "textDecorationLine": undefined,
            }
          }
        >
          Example Record
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        dataSet={{}}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={{}}
        testID="1234"
      >
        <View
          dataSet={{}}
          style={
            {
              "paddingBottom": 4,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 4,
            }
          }
        >
          <View
            dataSet={{}}
            style={{}}
          >
            <View
              dataSet={{}}
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <View
                dataSet={{}}
                style={
                  {
                    "width": "15%",
                  }
                }
              >
                <View
                  dataSet={{}}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#f0abfc",
                      "borderColor": "#27272a",
                      "borderRadius": 9999,
                      "height": 48,
                      "justifyContent": "center",
                      "marginTop": 4,
                      "position": "relative",
                      "width": 48,
                    }
                  }
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#fafafa",
                        "fontFamily": undefined,
                        "fontSize": 16,
                        "fontStyle": "normal",
                        "fontWeight": "600",
                        "letterSpacing": 0,
                        "lineHeight": 24,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    HI
                  </Text>
                </View>
              </View>
              <View
                dataSet={{}}
                style={
                  {
                    "paddingRight": 4,
                    "width": "78%",
                  }
                }
              >
                <View
                  dataSet={{}}
                  style={
                    {
                      "marginLeft": 8,
                    }
                  }
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#171717",
                        "fontFamily": undefined,
                        "fontSize": 14,
                        "fontStyle": "normal",
                        "fontWeight": "600",
                        "letterSpacing": 0,
                        "lineHeight": 21,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    Helpful Individual
                  </Text>
                  <View
                    dataSet={{}}
                    style={{}}
                  />
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#171717",
                        "fontFamily": undefined,
                        "fontSize": 14,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                        "lineHeight": 21,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    Community Scientist
                  </Text>
                  <View
                    dataSet={{}}
                    style={{}}
                  />
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#db2777",
                        "fontFamily": undefined,
                        "fontSize": 14,
                        "fontStyle": "normal",
                        "fontWeight": "700",
                        "letterSpacing": 0,
                        "lineHeight": 21,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    Sample, 2022-04-01
                  </Text>
                </View>
              </View>
              <View
                dataSet={{}}
                style={{}}
              />
              <View
                dataSet={{}}
                style={
                  {
                    "width": "7%",
                  }
                }
              />
            </View>
            <View
              dataSet={{}}
              style={{}}
            />
            <Text
              dataSet={{}}
              style={
                {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "500",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              White Chuck Glacier, Glacier Peak Wilderness, WA
            </Text>
            <View
              dataSet={{}}
              style={{}}
            />
            <Text
              dataSet={{}}
              style={
                {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Dark red algae in runnels
            </Text>
          </View>
        </View>
        <Image
          accessibilityLabel="static image"
          alt="static image"
          dataSet={{}}
          onError={[Function]}
          source={1}
          style={
            {
              "maxWidth": "100%",
            }
          }
        />
      </View>
      <View
        dataSet={{}}
        style={
          {
            "backgroundColor": "#d6d3d1",
            "height": 12,
            "width": "100%",
          }
        }
      />
    </View>
  </RCTScrollView>
</RNCSafeAreaProvider>
`;
